### DISCUSSION TITLE
v0.24.0 -> v0.25.0 Upgrade Notes for Release

### DISCUSSION THREAD

**Discussion by robacarp** [luckyframework/lucky#1359] [General] [answered]
This is a list of the issues I faced updating my app from lucky release 0.24.0 to master on lucky and avram.

Upgrade notes:

- `with_defaults` renamed to `tag_defaults` #1262 - easy search and replace 
- Avram::Operation now declares an abstract #run method which changes the interface a bit:
  - generated operations for ResetPasswordRequest and SignInUser need [to be updated](https://github.com/luckyframework/lucky/discussions/1359#discussioncomment-165907)
  - generated actions for PasswordResetRequests::Create and SignIns::Create need [to be updated](https://github.com/luckyframework/lucky/discussions/1359#discussioncomment-165923)
- `has_many through:` association declarations [need minor updating](https://github.com/luckyframework/avram/pull/525).
- `belongs_to` join-querying [need minor correcting](https://github.com/luckyframework/avram/pull/525)

Annoyances:
- Log spammed by "Executing query" debug messages [from Crystal DB](https://github.com/crystal-lang/crystal-db/blob/master/src/db/statement.cr#L145) - [update to generated code](https://github.com/luckyframework/lucky/discussions/1359#discussioncomment-216431) - [pull request](https://github.com/luckyframework/lucky_cli/pull/589/files)

Resolved:
- **resolved** by https://github.com/luckyframework/avram/pull/559. As of https://github.com/luckyframework/avram/pull/552 it's no longer by default supported to reference parameterized actions as `Action::Name.with(model_object)` -- instead try `Action::Name.with(model_object.id)`

**Comment by robacarp** [comment]
Update for ResetPasswordRequest, SignInUser is similar:

```diff

  # Run validations and yield the operation and the user if valid
-  def submit
+  def run
    user = user_from_email
    validate(user)

    if valid?
-      yield self, user
+      user
    else
-      yield self, nil
+      nil
    end
  end
```

**Comment by robacarp** [comment]
Update for PasswordResetRequests::Create, SignIns::Create is similar:

```diff
  route do
-    RequestPasswordReset.new(params).submit do |operation, user|
+    RequestPasswordReset.run(params) do |operation, user|
      if user
```

**Comment by robacarp** [comment]


**Reply by jwoertink** [reply]
This probably isn't necessary now since we added this back in https://github.com/luckyframework/avram/pull/559

**Reply by robacarp** [reply]
fantastic! I didn't see that go in.

**Comment by robacarp** [comment]
Generated code update for log "Executing Query" annoyance:

Update config/log.cr:
```diff
  # Use a pretty formatter printing to STDOUT in development
  backend = Log::IOBackend.new
  backend.formatter = Lucky::PrettyLogFormatter.proc
  Log.dexter.configure(:debug, backend)
+  DB::Log.level = :info
end
```

**Comment by stephendolan** [comment]
I've updated LuckyDiff as well:

https://luckydiff.com/?from=0.24.0&to=0.25.0

**Comment by jwoertink** [comment, marked as answer]
Thanks for putting this together @robacarp! It's all now ready for the UPGRADE_NOTES https://github.com/luckyframework/lucky/pull/1365/files Let me know if you find anything missing on there. That will be merged in maybe tomorrow.

**Reply by robacarp** [reply]
looks great, thanks!
